src/backend/input.rs

```diff
@@ -0,0 +1,277 @@
+/// be interpreted in the context of other events on the same slot.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct TouchSlot { id: u32 }
+
+impl NewIdType for TouchSlot
+{
+    fn new(id: u32) -> TouchSlot
+    {
+        TouchSlot { id: id }
+    }
+}
+
+/// Touch event
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum TouchEvent
+{
+    /// The start of an event at a given position (x, y).
+    ///
+    /// If the device has multi-touch capabilities a slot is given.
+    Down {
+        /// `TouchSlot`, if the device has multi-touch capabilities
+        slot: Option<TouchSlot>,
+        /// Absolute x-coordinate of the touch position.
+        x: f64,
+        /// Absolute y-coordinate of the touch position.
+        y: f64
+    },
+    /// Movement of a touch on the device surface to a given position (x, y).
+    ///
+    /// If the device has multi-touch capabilities a slot is given.
+    Motion {
+        /// `TouchSlot`, if the device has multi-touch capabilities
+        slot: Option<TouchSlot>,
+        /// Absolute x-coordinate of the final touch position after the motion.
+        x: f64,
+        /// Absolute y-coordinate of the final touch position after the motion.
+        y: f64 },
+    /// Stop of an event chain.
+    ///
+    /// If the device has multi-touch capabilities a slot is given.
+    Up {
+        /// `TouchSlot`, if the device has multi-touch capabilities
+        slot: Option<TouchSlot>
+    },
+    /// Cancel of an event chain. All previous events in the chain should be ignored.
+    ///
+    /// If the device has multi-touch capabilities a slot is given.
+    Cancel {
+        /// `TouchSlot`, if the device has multi-touch capabilities
+        slot: Option<TouchSlot>
+    },
+    /// Signals the end of a set of touchpoints at one device sample time.
+    Frame
+}
+
+/// Trait that describes objects providing a source of input events. All input backends
+/// need to implemenent this and provide the same base gurantees about the presicion of
+/// given events.
+pub trait InputBackend {
+    /// Sets a new handler for this `InputBackend`
+    fn set_handler<H: InputHandler + 'static>(&mut self, handler: H);
+    /// Get a reference to the currently set handler, if any
+    fn get_handler(&mut self) -> Option<&mut InputHandler>;
+    /// Clears the currently handler, if one is set
+    fn clear_handler(&mut self);
+    /// Sets the cursor position, useful for e.g. pointer wrapping.
+    ///
+    /// Not guaranteed to be supported on every backend. The result usually
+    /// depends on the capability of the backend, but may also fail for certain
+    /// specific actions. See the backends documentation.
+    fn set_cursor_position(&mut self, x: u32, y: u32) -> Result<(), ()>;
+}
+
+/// Implement to receive input events from any `InputBackend`.
+pub trait InputHandler {
+    /// Called when a new `Seat` has been created
+    fn on_seat_created(&mut self, seat: &Seat);
```

Comment:
I not sure this will to be an id type forever. E.g. even with a callback for capabilities, I still think it would make sense to expose the capabilities through `seat.capabilities()`.